### PR TITLE
Re: Abstract and Optimize Reclaim Labels

### PR DESCRIPTION
**Pull request by Hdt80bro**
Redirecting from https://github.com/4z0t/fa/pull/2 to here

### CODE REVIEW

File: lua/ui/game/reclaim.lua [hunk 1]
```diff
@@ -1,485 +1,499 @@
...
+            end
+        end
+    else
+        local x21, z21 = x2 - x1, z2 - z1
+        local x32, z32 = x3 - x2, z3 - z2
+        local x43, z43 = x4 - x3, z4 - z3
+        local x14, z14 = x1 - x4, z1 - z4
+
+        for _, recl in reclaim do
+            if recl.mass >= minAmount or true then
```
Comment (4z0t) [inline lua/ui/game/reclaim.lua:357]: Test code

File: lua/ui/game/reclaim.lua [hunk 2]
```diff
@@ -1,485 +1,499 @@
...
 
--- Creates an empty reclaim label
-function CreateReclaimLabel(view)
-    local label = WorldLabel(view, Vector(0, 0, 0))
-    return label
+---@param a UserReclaimData
+---@param b UserReclaimData
+---@return boolean
+local function CompareMass(a, b)
+    return a.mass < b.mass
```
Comment (4z0t) [inline lua/ui/game/reclaim.lua:275]: Wrong operator

Comment (Hdt80bro) [reply lua/ui/game/reclaim.lua:275]: Ah, I'm still based on an older fork. Good catch.

File: lua/ui/game/reclaim.lua [hunk 3]
```diff
@@ -1,485 +1,499 @@
...
+
+    local minX = MathMin(x1, x2, x3, x4)
+    local maxX = MathMax(x1, x2, x3, x4)
+    local minZ = MathMin(z1, z2, z3, z4)
+    local maxZ = MathMax(z1, z2, z3, z4)
+
+    if ContainsWholeMap(minX, minZ, maxX, maxZ) then
+        -- we can remove checking if in the prop is in view when we know all props are in view
+        for _, recl in reclaim do
+            if recl.mass >= minAmount or true then
```
[the remark this answers is not in the document]
Comment (Hdt80bro) [reply lua/ui/game/reclaim.lua:345]: Don't worry, it's still a draft--I didn't even resolve merge conflicts properly :)

File: lua/system/trashbag.lua [hunk 4]
```diff
@@ -24,6 +24,11 @@
 -- looking over at /lua/EffectTemplate.lua there can be up to nine emitters (oblivion_cannon_hit_08_emit)
 -- for a given effect.
 
+
+WeakKeyMetatable = {__mode = 'k'}
```
Comment (Garanas) [inline lua/system/trashbag.lua:28]: Good call 👍

File: lua/ui/game/reclaim.lua [hunk 5]
```diff
@@ -264,208 +237,514 @@ local WorldLabel = Class(Group) {
...
-        -- we can remove checking if in the prop is in view when we know all props are in view
+    -- The world view is an arbitrary quadrilateral, which means in order to see if a point is
+    -- inside it, we need to check if it is inside each of the four lines.
+    -- The equation to check if point (x, y) falls on the right of line AB is
+    --    (y - A.y) * (B.x - A.x) < (x - A.x) * (B.y - A.y)
+    -- which simplifies to
+    --    (A.y - B.y) * x + (B.x - A.x) * y < (A.y - B.y) * A.x + (B.x - A.x) * A.y
+    -- The view is actually in 3D space, but we ignore the Y component, so for us it's
+    --    (z1 - z2) * x + (x2 - x1) * z < (z1 - z2) * x1 + (x2 - x1) * z1
+    -- or, with the constant terms collapsed,
```
Comment (Garanas) [inline lua/ui/game/reclaim.lua:647]: Great documentation

**Comment by Garanas** [comment]
image

Looks exciting! Reporting an issue on my end

**Comment by Hdt80bro** [comment]
That should mitigate the problem for now, though I think the color scaling I put in should be adjusted.

File: lua/sim/Prop.lua [hunk 6]
```diff
@@ -294,9 +240,48 @@ Prop = Class(moho.prop_methods) {
...
 
-        -- Notify UI about the mass change
-        self:SyncMassLabel()
-    end,
+        -- Notify UI about the change
+        self:SyncUserData()
+    end;
+
+    --- Syncs the reclaim label to the UI
+    SyncUserData = function(self)
```
[the remark this answers is not in the document]
Comment (Hdt80bro) [reply lua/sim/Prop.lua:248]: I tried to leave renaming alone, even when it was painful (ahem, *OutsidePlayableAreaReclaim*), but I don't imagine anyone is manually syncing prop reclaim values. Do let me know if you find something to the contrary, I know you have a large mod library.

**Comment by 4z0t** [comment]
I discourage this kind of PRs, lots of changes not related to its name.

**Comment by Hdt80bro** [comment]
Nearly all of it is--the only thing I could have left out were some of the Layout helpers, but I already needed some of changes and the rest needed to be done as well. Not worth a separate PR. You might be looking at some of the changes that affect other files (the annotation stuff in particular can look unrelated), however, those are things that *shouldn't* be done separately. We have tried separating out aspects into different PR's in the past and it is nasty.

**Comment by Garanas** [comment]
The performance degraded quite a bit, in comparison to what we have in `deploy/fafdevelop`, do you have an idea where the extra time is going towards to?

**Comment by Hdt80bro** [comment]
Yes, I made the mistake of optimizing *before* abstracting... it's finally working like I want it to, but I need to do another optimization pass. I suspect it's how I'm handling frame updates.

File: lua/ScenarioFramework.lua [hunk 7]
```diff
@@ -1677,10 +1677,10 @@ end
 
 --- Returns a random entry from a table
 ---@generic T
----@param table table<T>
+---@param tbl table<T>
 ---@return T
-function GetRandomEntry(table)
-    return table[Random(1, table.getn(table))]
+function GetRandomEntry(tbl)
+    return table[Random(1, table.getn(tbl))]
```
Comment (Garanas) [inline lua/ScenarioFramework.lua:1683]: shoudn't `table` here be `tbl` too (the lexically first one)

File: lua/lazyvar.lua [hunk 8]
```diff
@@ -2,21 +2,21 @@
...
 local setmetatable = setmetatable
 
 
+---@alias Lazy<T> T | LazyVar<T> | fun(): T
+
+
 -- Set this true to get tracebacks in error messages. It slows down lazyvars a lot,
 -- so don't use except when debugging.
-ExtendedErrorMessages = false
+ExtendedErrorMessages = true
```
Comment (Garanas) [inline lua/lazyvar.lua:17]: Should be set to false again

File: lua/sim/Prop.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (Garanas) [reply lua/sim/Prop.lua:248]: This appears to be save

**Comment by Garanas** [comment]
I've done a review and I'm not sure. 

When moving the camera this solution lags more, causing it to dip down to 20 - 30 fps on my computer. Meanwhile the solution we have on `deploy/fafdevelop` remains at 60 fps or more. I like a lot of the work in this pull request, one of them being the smoother colors. But I'm not sure if the performance drop is acceptable, especially as I have a relatively fast computer.

Let us sit down together next week and discuss it over voice.

**Comment by 4z0t** [comment]
How i understand you xd

**Comment by Garanas** [comment]
@Hdt80bro I'll close this pull request tomorrow. Time caught up and this pull request feels dated at this point

**Pull request closed without merging**